File: Chapter04/Exercise23_04/Exercise23_04.py
from pyspark.rdd import RDD

from Chapter01.utilities01_py.helper_python import create_session

if __name__ == "__main__":
    # 23 a
    from typing import List
    million_ints: List[int] = []
    for number in range(0, 1000000):
        million_ints.append(-1)

    from pympler import asizeof
    print(asizeof.asizeof(million_ints))
    #
    from Chapter02.utilities02_py.domain_objects  import WarcRecord
    million_warcs = list()
    for index in range(0, 1000000):
        million_warcs.append(WarcRecord.create_dummy())
    print(asizeof.asizeof(million_warcs))

    ##########################################################################
    spark = create_session(2, "Collection Sizes")
    ## (B)

    million_ints_rdd = spark.sparkContext.range(0, 1000000)
    million_ints_rdd.cache()
    million_ints_rdd.count()

    million_ints_rdd.unpersist()
    million_warcs_rdd = million_ints_rdd.map(lambda number: WarcRecord.create_dummy())
    million_warcs_rdd.cache()
    million_warcs_rdd.count()

    from pyspark.sql import DataFrame
    million_ints_df: DataFrame = spark.range(0, 1000000)
    million_ints_df.cache()
    million_ints_df.count()

    million_ints_df.unpersist()
    million_warcs_df: DataFrame = spark.createDataFrame(million_warcs)
    million_warcs_df.cache()
    million_warcs_df.count()

File: Chapter04/Exercise23_04/Exercise23_04.scala
package Exercise23_04

import Utilities01.HelperScala.createSession

object Exercise21_04 {
  def main(args: Array[String]): Unit = {
    // 23 a
    val millionInts: List[Int] = List.fill(1000000)(-1)
    val millionLongs: List[Long] = List.fill(1000000)(-1L)

    import org.apache.spark.util.SizeEstimator
    println(SizeEstimator.estimate(millionInts))
    println(SizeEstimator.estimate(millionLongs))

    import Utilities02.WarcRecord
    val millionWarcs: List[WarcRecord] = millionInts.map(_ => WarcRecord.createDummyObject())
    println(SizeEstimator.estimate(millionWarcs))

    /////////////////////////////////////////////////////////////////////////////////////
    val spark = createSession(2, "UI Exploration")
    // B

    import org.apache.spark.rdd.RDD
    val millionLongsRDD: RDD[Long] = spark.sparkContext.range(0, 1000000)
    millionLongsRDD.cache()
    millionLongsRDD.count()

    millionLongsRDD.unpersist()
    val millionWarcsRDD: RDD[WarcRecord] = millionLongsRDD.map(_ => WarcRecord.createDummyObject())
    millionWarcsRDD.cache()
    millionWarcsRDD.count()

    // DataFrames
    millionWarcsRDD.unpersist()
    import org.apache.spark.sql.DataFrame
    import spark.implicits._
    val millionLongsDF: DataFrame = millionLongsRDD.toDF()
    millionLongsDF.cache()
    millionLongsDF.count()

    millionLongsDF.unpersist()
    val millionWarcsDF: DataFrame = millionWarcsRDD.toDF()
    millionWarcsDF.cache()
    millionWarcsDF.count()

    // Datasets:
    millionWarcsDF.unpersist()
    import org.apache.spark.sql.Dataset
    val millionLongsDS: Dataset[Long] = millionLongsRDD.toDS()
    millionLongsDS.cache()
    millionLongsDS.count()

    millionLongsDS.unpersist()
    val millionWarcsDS: Dataset[WarcRecord] = millionWarcsRDD.toDS()
    millionWarcsDS.cache()
    millionWarcsDS.count()

    millionWarcsDS.unpersist()

  }
}